cm31_ntt/src/ntt_utils.rs

```diff
@@ -0,0 +1,260 @@
+use crate::cm31::{ CF, gen_roots_of_unity, W_8 };
+use crate::rm31::RF;
+use num_traits::{Zero, One};
+use num_traits::pow::Pow;
+use rayon::prelude::*;
+
+pub fn is_power_of(n: u32, x: u32) -> bool {
+    if n == 0 {
+        return false;
+    }
+
+    if x == 1 && n == 1 {
+        return true;
+    }
+
+    if x == 1 && n != 1 {
+        return false;
+    }
+
+    let mut num = n;
+    while num % x == 0 {
+        num /= x;
+    }
+
+    num == 1
+}
+
+
+/// Helper function to get the n-th root of unity in CF.
+pub fn get_root_of_unity(n: usize) -> CF {
+    assert!(n.is_power_of_two(), "n must be a power of 2");
+    let roots_of_unity = gen_roots_of_unity((n as f64).log2() as usize);
+    roots_of_unity[roots_of_unity.len() - 1]
+}
+
+/// A radix-8 NTT butterfly.
+#[inline]
+pub fn ntt_block_8(
+    f0: CF,
+    f1: CF,
+    f2: CF,
+    f3: CF,
+    f4: CF,
+    f5: CF,
+    f6: CF,
+    f7: CF,
+    wt: CF,
+    wt2: CF,
+    wt3: CF,
+    wt4: CF,
+    wt5: CF,
+    wt6: CF,
+    wt7: CF,
+) -> (CF, CF, CF, CF, CF, CF, CF, CF) {
+    // Refer to Yuval's Radix 8 DIT diagram.
+    // 1st columm of black dots: a0-a8
+    // 2nd columm of black dots: b0-b8
+    // 3nd columm of black dots: res[0]-res[8]
+
+    let t0 = f0;
+    let t1 = f1 * wt;
+    let t2 = f2 * wt2;
+    let t3 = f3 * wt3;
+    let t4 = f4 * wt4;
+    let t5 = f5 * wt5;
+    let t6 = f6 * wt6;
+    let t7 = f7 * wt7;
+
+    // Column 1
+    let a0 = t0 + t4;
+    let a1 = t0 - t4;
+    let a2 = t2 + t6;
+    let a3 = t2 - t6;
+    let a4 = t1 + t5;
+    let a5 = t1 - t5;
+    let a6 = t3 + t7;
+    let a7 = t3 - t7;
+
+    // Column 2
+    let a3_j = a3.mul_j();
+    let a7_j = a7.mul_j();
+
+    let b0 = a0 + a2;
+    let b1 = a0 - a2;
+    let b2 = a1 + a3_j;
+    let b3 = a1 - a3_j;
+    let b4 = a4 + a6;
+    let b5 = a4 - a6;
+    let b6 = a5 + a7_j;
+    let b7 = a5 - a7_j;
+
+    // Column 3
+    let b5_j = b5.mul_j();
+    let b7_j = b7.mul_j();
+    let b6_w8 = b6 * W_8;
+    let b7_j_w8 = b7_j * W_8;
```

Comment:
I'm surprised we are not exploiting the special structure here. Does the compiler turn this into bitshifts?

Reply:
Resolved in commit 496dbce0, though benchmarks don't show a significant improvement.